language/stdlib/modules/option.move

```diff
@@ -0,0 +1,119 @@
+module Option {
+    use 0x0::Vector;
+
+    // Abstraction of a value that may or may not be present. Implemented with a vector of size
+    // zero or one because Move bytecode does not have ADTs.
+    struct T<Element> { vec: vector<Element> }
+
+    // Return an empty `Option::T`
+    public fun none<Element>(): T<Element> {
+        T { vec: Vector::empty() }
+    }
+
+    // Return an `Option::T` containing `e`
+    public fun some<Element>(e: Element): T<Element> {
+        T { vec: Vector::singleton(e) }
+    }
+
+    // Return true if `t` does not hold a value
+    public fun is_none<Element>(t: &T<Element>): bool {
+        Vector::is_empty(&t.vec)
+    }
+
+    // Return true if `t` holds a value
+    public fun is_some<Element>(t: &T<Element>): bool {
+        !Vector::is_empty(&t.vec)
+    }
+
+    // Return true if the value in `t` is equal to `e_ref`
+    // Always returns `false` if `t` does not hold a value
+    public fun contains<Element>(t: &T<Element>, e_ref: &Element): bool {
+        Vector::borrow(&t.vec, 0) == e_ref
+    }
+
+    // Return an immutable reference to the value inside `t`
+    // Aborts if `t` does not hold a value
+    public fun borrow<Element>(t: &T<Element>): &Element {
+        Vector::borrow(&t.vec, 0)
+    }
+
+    // TODO: not clear this API is a good idea
+    // Return a reference to the value inside `t` if it holds one
+    // Return `default_ref` if `t` does not hold a value
+    public fun borrow_default<Element>(t: &T<Element>, default_ref: &Element): &Element {
+        let vec_ref = &t.vec;
+        if (Vector::is_empty(vec_ref)) {
```

Comment:
On `is_some` vs using the vector functions directly: I was trying to avoid the overhead of an extra call. In the long run, maybe we can do something like allowing you to write `is_some` (which is definitely clearer) and giving the compiler a hint to inline or letting it decide on its own...

Reply:
I feel like avoiding proper encapsulation for an extra call feels like a bad idea. I wouldn't worry too much about the fine grained gas costings of these functions (as we don't have custom code publishing). To me, we should prioritize readability and correctness over gas cost. Which I think would mean developing a small set of "core" functions that deal with the vector field directly, everything else is defined in terms of those functions. 

In short, I really think you should use `is_some` here. I don't think the brittleness is worth the gas cost.

Reply:
I agree with the general design principle or writing a set of core functions that manipulate the richer API (e.g., vector) and  then writing everything in terms of those. But: 
- I don't think it makes sense to write a core function for `Vector::borrow` (since it would be a wrapper that does exactly the same thing)
- I personally find it less readable to mix locally defined functions like `is_some` with raw accesses to `Vector` than to use all raw accesses

I'm happy to change this if it's a sticking point  or if I'm missing some cleaner way to write this in terms of core functions with no `Vector` accesses at all.